### PR TITLE
Add TOF hits into tracking

### PR DESCRIPTION
**Pull request by yezhenyu2003**
Add TOF as a TrackerSubAssembly_4 into tracker.xml

### CODE REVIEW

**Comment by sly2j** [comment]
@yezhenyu2003 to add a "null" tracker needed to disable the negative endocarp, you can create the following type of empty detector, and add it to the assembly:
https://eicweb.phy.anl.gov/EIC/detectors/athena/-/blob/master/compact/tracking_config_canyonlands.xml#L78
(replace `athena` with `epic`).

**Comment by yezhenyu2003** [comment]
> @yezhenyu2003 to add a "null" tracker needed to disable the negative endcap, you can create the following type of empty detector, and add it to the assembly:
> https://eicweb.phy.anl.gov/EIC/detectors/athena/-/blob/master/compact/tracking_config_canyonlands.xml#L78
> (replace `athena` with `epic`).

Added a null tracker (BackwardTOF) into tof_endcap.xml, but tracker_geometry check is failing

D2A_V:Endcap   ERROR     Misconfiguration in volume building detected.

**Comment by yezhenyu2003** [comment]
Added a null tracker (BackwardTOF) into tof_endcap.xml, but tracker_geometry check is failing

D2A_V:Endcap ERROR Misconfiguration in volume building detected.

**Review by wdconinc — COMMENTED** [review]
Can you address why `BarrelTOF_radius_design = 63.0 cm` becomes `BarrelTOF_radius = 63.0 cm` becomes `BarrelTOF_rmin = 63.0 cm` but then (with `BarrelTOF_rOffset = 16.5*mm`)
```xml
        <barrel_envelope
          inner_r="BarrelTOF_rmin-BarrelTOF_rOffset"
          outer_r="BarrelTOF_rmax+BarrelTOF_rOffset"
```
where the envelope is now outside the 63.0 cm to 66.0 cm range in the detector geometry database, https://eic.jlab.org/Geometry/Detector/Detector-20220929172703.html.

**Comment by yezhenyu2003** [comment]
> Can you address why `BarrelTOF_radius_design = 63.0 cm` becomes `BarrelTOF_radius = 63.0 cm` becomes `BarrelTOF_rmin = 63.0 cm` but then (with `BarrelTOF_rOffset = 16.5*mm`)
> 
> ```
>         <barrel_envelope
>           inner_r="BarrelTOF_rmin-BarrelTOF_rOffset"
>           outer_r="BarrelTOF_rmax+BarrelTOF_rOffset"
> ```
> 
> where the envelope is now outside the 63.0 cm to 66.0 cm range in the detector geometry database, https://eic.jlab.org/Geometry/Detector/Detector-20220929172703.html.

Fixed in tof_barrrel.xml

**Comment by osbornjd** [comment]
As discussed today, this should be ready to go

**Comment by yezhenyu2003** [comment]
> As discussed today, this should be ready to go

Yes, waiting for review/approval

File: compact/tof_barrel.xml [hunk 1]
```diff
@@ -107,18 +115,18 @@
       </module>
       <layer module="BarrelTOF_Module1" id="1" vis="TOFLayerVis">
         <barrel_envelope
-          inner_r="BarrelTOF_rmin-BarrelTOF_rOffset"
-          outer_r="BarrelTOF_rmax+BarrelTOF_rOffset"
-          z_length="BarrelTOF_length"/>
-        <rphi_layout phi_tilt="BarrelTOF_Module_tiltangle" nphi="BarrelTOF_Module_nphi" phi0="0.0" rc="BarrelTOF_rmin" dr="0.0*mm"/>
-        <z_layout dr="0.0*mm" z0="15.0*cm" nz="BarrelTOF_Module_nz"/>
+          inner_r="BarrelTOF_rmin-1*mm"
```
Comment (wdconinc) [inline compact/tof_barrel.xml:118]: You cannot go less than `rmin` at all, not even a millimeter. That will(!) result in conflicts with other volumes.

File: configurations/full.yml [hunk 2]
```diff
@@ -2,8 +2,8 @@ features:
   beampipe:
   tracking:
   pid:
-    tof_endcap:
-    tof_barrel:
+    #tof_endcap:
+    #tof_barrel:
```
Comment (wdconinc) [inline configurations/full.yml:6]: No commented out parts, please. Just remove.

File: compact/tracker.xml [hunk 3]
```diff
@@ -119,12 +119,25 @@
...
       vis="TrackerSubAssemblyVis">
       <composite name="OuterMPGDBarrel"/>
     </detector>
+    <detector id="TrackerSubAssembly_6_ID"
+      name="EndcapTOFSubSubAssembly"
+      type="DD4hep_SubdetectorAssembly"
+      vis="TrackerSubAssemblyVis">
+      <composite name="ForwardTOF"/>
+      <composite name="BackwardTOF"/>
+    </detector>
```
Comment (wdconinc) [inline compact/tracker.xml:140]: This has overlap with #191 where it is reported as failing by Nicolas. @sly2j was going to look into that, and should review here.

File: compact/tof_endcap.xml [hunk 4]
```diff
@@ -130,12 +130,50 @@
...
+      </detector>
+      <detector id="BackwardTOF_ID"
+        name="BackwardTOF"
+        type="epic_CompositeTracker"
+        actsType="endcap"
+        vis="TrackerSubAssemblyVis">
+        <position x="0*cm" y="0*cm" z="-1*um" />
+      </detector>
+
+      <!--detector
```
Comment (wdconinc) [inline compact/tof_endcap.xml:176]: Do not leave large blocks of commented geometry in the files. Just remove. (git has a history for that reason)

File: compact/tof_barrel.xml (hunk 1 above)
Comment (yezhenyu2003) [reply compact/tof_barrel.xml]: Fixed

File: configurations/full.yml (hunk 2 above)
Comment (yezhenyu2003) [reply configurations/full.yml]: Fixed

File: compact/tof_endcap.xml (hunk 4 above)
Comment (yezhenyu2003) [reply compact/tof_endcap.xml]: Fixed

File: compact/tracker.xml (hunk 3 above)
Comment (yezhenyu2003) [reply compact/tracker.xml:140]: Yes. The recommendation from Tuesday's CompSW meeting by Joe and Sylvester is to move forward with merging this into main branch, while investigating into #191.

**Review by wdconinc — DISMISSED** [review]

**Comment by wdconinc** [comment]
This should merge as soon as checks complete.

**Comment by wdconinc** [comment]
Cannot auto-update because of conflicts.

**Comment by yezhenyu2003** [comment]
What conflicts?

**Comment by wdconinc** [comment]
That's an automated message. I'd have to look...

**Comment by wdconinc** [comment]
There is a lot of stuff going into the main branch and many things are getting updated, sometimes in the same areas of the code. Take a look at the pull request and it will show you.

**Comment by yezhenyu2003** [comment]
I updated the files to resolve the conflicts. The remaining differences are intentional for TOF.

Zhenyu

**Review by wdconinc — APPROVED** [review]

File: compact/definitions.xml [hunk 5]
```diff
@@ -442,7 +443,7 @@ Examples:
     <comment> Global PID regions with suballocations for TOF and RICH detectors </comment>
     <constant name="ForwardPIDRegion_zmin"        value="CentralTrackingRegionP_zmax" />
     <constant name="ForwardPIDRegion_length"      value="142.0*cm" />
-    <constant name="ForwardPIDRegion_rmax"        value="185.0*cm" />
+    <constant name="ForwardPIDRegion_rmax"        value="180.0*cm" />
```
Comment (wdconinc) [inline compact/definitions.xml:446]: Related to PR?

File: compact/definitions.xml [hunk 6]
```diff
@@ -502,15 +503,14 @@ Service gaps in FW direction (before endcapP ECAL) and BW direction (before endc
     <constant name="ForwardServiceGap_zmax"       value="ForwardServiceGap_zmin + ForwardServiceGap_length"/>
 
     <comment> Radial service gap between dRICH and HcalBarrel </comment>
-    <constant name="ForwardServiceGap_thickness"  value="10.0*cm"/>
+    <constant name="ForwardServiceGap_thickness"  value="15.0*cm"/>
```
Comment (wdconinc) [inline compact/definitions.xml:506]: Related to PR?

File: compact/definitions.xml [hunk 7]
```diff
@@ -535,9 +535,8 @@ Service gaps in FW direction (before endcapP ECAL) and BW direction (before endc
     <constant name="EcalEndcapPInsert_hole_xposition" value="-7.29*cm"/>
     <constant name="EcalEndcapPInsert_hole_yposition" value="0.0*cm"/>
 
-    <!-- <constant name="EcalEndcapN_zmin"               value="BackwardPIDRegion_zmin + BackwardInnerEndcapRegion_length"/> -->
-    <constant name="EcalEndcapN_zmin"               value="174*cm"/>   <!-- Currently fix value hardcoded -->
-    <constant name="EcalEndcapN_length"             value="60*cm" />   <!-- Currently fix value hardcoded -->
+    <constant name="EcalEndcapN_zmin"               value="BackwardPIDRegion_zmin + BackwardInnerEndcapRegion_length"/>
+    <constant name="EcalEndcapN_length"             value="60*cm" />
```
Comment (wdconinc) [inline compact/definitions.xml:539]: Related to PR?

File: compact/definitions.xml [hunk 8]
```diff
@@ -546,8 +545,8 @@ Service gaps in FW direction (before endcapP ECAL) and BW direction (before endc
     </comment>
     <constant name="EcalEndcapN_rmin1"              value="Eta4_6_tan * EcalEndcapN_zmin"/>
     <constant name="EcalEndcapN_rmin2"              value="Eta3_9_tan * EcalEndcapN_zmin"/>
-    <constant name="EcalEndcapN_rmin"               value="9.*cm"/>    <!-- Currently fix value hardcoded -->
-    <constant name="EcalEndcapN_rmax"               value="63.*cm"/>   <!-- Currently fix value hardcoded -->
+    <constant name="EcalEndcapN_rmin"               value="max(EcalEndcapN_rmin1, EcalEndcapN_rmin2)"/>
+    <constant name="EcalEndcapN_rmax"               value="63*cm"/> <!-- FIXME hardcoded -->
```
Comment (wdconinc) [inline compact/definitions.xml:549]: Related to PR?

**Review by wdconinc — CHANGES_REQUESTED** [review]

File: compact/definitions.xml (hunk 5 above)
Comment (yezhenyu2003) [reply compact/definitions.xml]: fixed

File: compact/definitions.xml (hunk 6 above)
Comment (yezhenyu2003) [reply compact/definitions.xml]: fixed

File: compact/definitions.xml (hunk 7 above)
Comment (yezhenyu2003) [reply compact/definitions.xml]: fixed

File: compact/definitions.xml (hunk 8 above)
Comment (yezhenyu2003) [reply compact/definitions.xml]: fixed

**Review by wdconinc — APPROVED** [review]

**Pull request merged**